okhttp/src/main/java/okhttp3/Challenge.kt

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (C) 2019 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3
+
+import java.nio.charset.Charset
+import java.nio.charset.StandardCharsets.ISO_8859_1
+import java.util.Collections.singletonMap
+import java.util.Collections.unmodifiableMap
+import java.util.Locale.US
+
+/** An RFC 7235 challenge. */
+class Challenge(
+  private val scheme: String,
+  authParams: Map<String?, String>
+) {
+  private val authParams: Map<String?, String>
+
+  init {
+    val newAuthParams: MutableMap<String?, String> = LinkedHashMap()
+    for ((key, value) in authParams) {
+      val newKey = key?.toLowerCase(US)
```

Comment:
In the Java code, we check for whether the `key` is `null` and so I'm carrying this over. This does mean that the `authParams` are `Map<String?, String>`.

Reply:
All maps coming from Java may have a null key. Whether we actually specifically allow that here I'm not sure.

Reply:
We describe it as the key used for token68.